cpp/modmesh/buffer/SimpleArray.cpp

```diff
@@ -28,56 +28,46 @@
 
 #include <modmesh/buffer/SimpleArray.hpp>
 
+#include <unordered_map>
+
 namespace modmesh
 {
 
-DataType get_data_type_from_string(const std::string & data_type_string)
+namespace detail
 {
-    if (data_type_string == "bool")
-    {
-        return DataType::Bool;
-    }
-    if (data_type_string == "int8")
-    {
-        return DataType::Int8;
-    }
-    if (data_type_string == "int16")
-    {
-        return DataType::Int16;
-    }
-    if (data_type_string == "int32")
-    {
-        return DataType::Int32;
-    }
-    if (data_type_string == "int64")
-    {
-        return DataType::Uint64;
-    }
-    if (data_type_string == "uint8")
-    {
-        return DataType::Uint8;
-    }
-    if (data_type_string == "uint16")
-    {
-        return DataType::Uint16;
-    }
-    if (data_type_string == "uint32")
-    {
-        return DataType::Uint32;
-    }
-    if (data_type_string == "uint64")
-    {
-        return DataType::Uint64;
-    }
-    if (data_type_string == "float32")
-    {
-        return DataType::Float32;
-    }
-    if (data_type_string == "float64")
+struct DataTypeHasher
+{
+    std::size_t operator()(const std::string & data_type_string) const
+    {
+        const char ch1 = data_type_string.front();
+        const char ch2 = data_type_string.back();
+        return static_cast<std::size_t>(ch1) << 8 | static_cast<std::size_t>(ch2);
+    }
+};
+
+// NOLINTNEXTLINE(misc-use-anonymous-namespace, cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables, fuchsia-statically-constructed-objects)
+static std::unordered_map<std::string, DataType, DataTypeHasher> string_data_type_map = {
+    {"bool", DataType::Bool},
+    {"int8", DataType::Int8},
+    {"int16", DataType::Int16},
+    {"int32", DataType::Int32},
+    {"int64", DataType::Int64},
+    {"uint8", DataType::Uint8},
+    {"uint16", DataType::Uint16},
+    {"uint32", DataType::Uint32},
+    {"uint64", DataType::Uint64},
+    {"float32", DataType::Float32},
+    {"float64", DataType::Float64}};
+} // end of namespace detail
+
+DataType::DataType(const std::string & data_type_string)
```

Comment:
Could you check if `inline` makes more efficient assembly?

Reply:
I tried, but the compiler didn't allow me to put "inline" here. There is a linking issue if I add `inline`. The signature should be the line(?)

if I added `inline`, error:
```
[ 82%] Linking CXX executable test_nopython
/usr/bin/ld: CMakeFiles/test_nopython.dir/test_nopython_buffer.cpp.o: in function `SimpleArray_DataType_from_string_Test::TestBody()':
/mnt/c/Users/tiger/modmesh/gtests/test_nopython_buffer.cpp:78: undefined reference to `modmesh::DataType::DataType(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /mnt/c/Users/tiger/modmesh/gtests/test_nopython_buffer.cpp:81: undefined reference to `modmesh::DataType::DataType(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /mnt/c/Users/tiger/modmesh/gtests/test_nopython_buffer.cpp:84: undefined reference to `modmesh::DataType::DataType(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /mnt/c/Users/tiger/modmesh/gtests/test_nopython_buffer.cpp:85: undefined reference to `modmesh::DataType::DataType(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```

Reply:
so I guess we need to put the implementation in hpp to avoid this issue; however, I think this implementation is better to keep in cpp, because it also use other static object in `modmesh::detail` in cpp.

Reply:
You are right.  The whole things need to be in header files.  Let's try it later.